File: fake-service/features/step_definitions/mail_list.rb
When(/^I open the first mail in the '(.*)'$/) do |tag|
  page.execute_script("window.scrollBy(0, -200)")
  step "I select the tag '#{tag}'"
  step 'I open the first mail in the mail list'
end

When(/^I open the first mail in the mail list$/) do
  within('#mail-list') do
    mail_link = first('a')
    @current_mail_id = mail_link.native.attribute('href').scan(/\/(\d+)$/).flatten.first
    begin
      mail_link.click
    rescue # in Chrome, the 'a' in mail_list is not clickable because it's hidden inside the 'li'
      mail_link_parent_li = mail_link.find(:xpath, '../..')
      mail_link_parent_li.click
    end
  end
end

When(/I see that mail under the '(.*)' tag/) do |tag|
  step "I select the tag '#{tag}'"
  check_current_mail_is_visible
end

And(/^I open the mail I previously tagged$/) do
  open_current_mail
end

When(/^I open that mail$/) do
  open_current_mail
end

Then(/^I see the mail I sent$/) do
  check_current_mail_is_visible
end

Then(/^the deleted mail is there$/) do
  check_current_mail_is_visible
end

def open_current_mail
  within('#mail-list') do
    begin
      first("#mail-#{@current_mail_id} a").click
    rescue # in Chrome, the 'a' in mail_list is not clickable because it's hidden inside the 'li'
      first("#mail-#{@current_mail_id}").click
    end
  end
end

def check_current_mail_is_visible
  within('#mail-list') do
    have_selector?("#mail-#{@current_mail_id}").should be_true
  end
end

File: py-fake-service/features/steps/mail_list.py
import re
from behave import *
from common import *

def find_current_mail(context):
    return find_element_by_xpath(context, '//*[@id="mail-list"]/li[@id="mail-%s"]//a' % context.current_mail_id)


def check_current_mail_is_visible(context):
    find_current_mail(context)

def open_current_mail(context):
    e = find_current_mail(context)
    e.click()

@then('I see that mail under the \'{tag}\' tag')
def impl(context, tag):
    context.execute_steps("when I select the tag '%s'" % tag)
    check_current_mail_is_visible(context)

@when('I open that mail')
def impl(context):
	open_current_mail(context)

@when('I open the first mail in the mail list')
def impl(context):
    elements = context.browser.find_elements_by_xpath('//*[@id="mail-list"]//a')
    context.current_mail_id = elements[0].get_attribute('href').split('/')[-1]
    elements[0].click()

@when('I open the first mail in the \'{tag}\'')
def impl(context, tag):
    context.browser.execute_script('window.scrollBy(0, -200)')
    context.execute_steps(u"When I select the tag '%s'" % tag)
    context.execute_steps(u'When I open the first mail in the mail list')

@then('I open the mail I previously tagged')
def impl(context):
    open_current_mail(context)

@then('I see the mail I sent')
def impl(context):
    src = context.browser.page_source
    assert_that(src, contains_string(context.reply_subject))

@then('the deleted mail is there')
def impl(context):
    check_current_mail_is_visible(context)

